src/read/macho/file.rs

```diff
@@ -287,32 +287,26 @@ where
         &'file self,
         section_name: &[u8],
     ) -> Option<MachOSection<'data, 'file, Mach, R>> {
-        // Translate the "." prefix to the "__" prefix used by OSX/Mach-O, eg
-        // ".debug_info" to "__debug_info", and limit to 16 bytes total.
-        let system_name = if section_name.starts_with(b".") {
-            if section_name.len() > 15 {
-                Some(&section_name[1..15])
-            } else {
-                Some(&section_name[1..])
-            }
-        } else {
-            None
+        // Compute the portion of the section name after the "." prefix in order
+        // to match against the "__" and "__z" prefixes.
+        let base_name = section_name.strip_prefix(b".");
+        // Section names in Mach-O files are limited to 16 characters. If the
+        // queried section_name is longer, we want to match only the portion
+        // following the prefix up to that limit.
+        const MAX_SECTION_NAME_LEN: usize = 16;
+        let make_suffix = |prefix: &[u8]| {
+            let max_len = MAX_SECTION_NAME_LEN - prefix.len();
+            base_name.map(|name| name.get(..max_len).unwrap_or(name))
         };
-        let cmp_section_name = |section: &MachOSection<'data, 'file, Mach, R>| {
-            section
-                .name_bytes()
-                .map(|name| {
-                    section_name == name
-                        || system_name
-                            .filter(|system_name| {
-                                name.starts_with(b"__") && name[2..] == **system_name
-                            })
-                            .is_some()
-                })
-                .unwrap_or(false)
+        let underscores_suffix = make_suffix(b"__");
+        let underscores_z_suffix = make_suffix(b"__z");
+        let name_matches = |name: &[u8]| {
+            name == section_name
+                || name.strip_prefix(b"__") == underscores_suffix
+                || name.strip_prefix(b"__z") == underscores_z_suffix
```

Comment:
This code looks a lot nicer than what we had before. However, these two comparisons will incorrectly match if both sides are `None` (e.g. this `__z` check will result in matching `__text` with `__data`).

Reply:
Good catch! I've found a formulation I like more than before that handles this case. See how you feel about it.